docker/Dockerfile.local

```diff
@@ -1,34 +1,76 @@
-FROM nvidia/cuda:11.3.1-devel-ubuntu20.04
+ARG DEVICE=gpu
+ARG SOURCE
+
+FROM ${SOURCE} as base
+
 ENV DEBIAN_FRONTEND=noninteractive
 ENV HOME=/root
 
-RUN apt-get update
-RUN apt-get install -y --no-install-recommends build-essential python3-dev make g++ vim net-tools
+ARG DGL_VERSION=1.1.3
+ARG OGB_VERSION=1.3.6
+ARG TORCH_VERSION=2.1.2
+ARG TRANSFORMERS_VERSION=4.28.1
 
-RUN apt-get install -y python3-pip git wget psmisc
-RUN apt-get install -y cmake
+RUN apt update && apt install -y --no-install-recommends \
+    git \
+    libicu-dev \
+    openssh-client \
+    openssh-server \
+    python3.9 \
+    python3.9-distutils \
+    python3.9-venv \
+    && rm -rf /var/lib/apt/lists/*
 
-# Install Pytorch
-RUN pip3 install networkx==3.1 pydantic
-RUN pip3 install torch==2.1.0+cu118 --extra-index-url https://download.pytorch.org/whl/cu118
+# Create and activate a Python venv
+RUN python3.9 -m venv /opt/gs-venv
+ENV PATH="/opt/gs-venv/bin:$PATH"
 
-# Install DGL
-RUN pip3 install dgl==1.0.4+cu117 -f https://data.dgl.ai/wheels/cu117/repo.html
-ENV PYTHONPATH="/root/dgl/tools/:${PYTHONPATH}"
+# Install GraphStorm dependencies
+RUN pip install \
+    boto3==1.34.89 \
+    botocore==1.34.89 \
+    h5py==3.11.0 \
+    networkx==3.1 \
+    psutil==5.9.8 \
+    pyarrow==14.0.0 \
+    pydantic==2.7.0 \
+    scikit-learn==1.4.2 \
+    scipy==1.13.0 \
+    && rm -rf /root/.cache
+
+FROM base as base-cpu
+
+# Install torch, DGL, and GSF deps that require torch
+RUN pip install \
+    torch==${TORCH_VERSION} \
+    --index-url https://download.pytorch.org/whl/cpu \
+    && rm -rf /root/.cache
 
-# Install related Python packages
-RUN pip3 install ogb==1.3.6 scipy pyarrow boto3 scikit-learn transformers
+RUN pip install \
+    dgl==${DGL_VERSION} \
+    ogb==${OGB_VERSION} \
+    transformers==${TRANSFORMERS_VERSION} \
+    -f https://data.dgl.ai/wheels-internal/repo.html && rm -rf /root/.cache
 
-# Install other dependencies
-RUN apt-get install -y cython3 libicu-dev
-RUN pip3 install h5py psutil
+FROM base as base-gpu
 
-RUN apt-get install -y unzip
+# Install torch, DGL, and GSF deps that require torch
+RUN pip install \
+    dgl==${DGL_VERSION}+cu121 \
+    ogb==${OGB_VERSION} \
+    torch==${TORCH_VERSION} \
```

Comment:
Why do we not specify the torch version with CUDA version any more here? I see we are doing so for dgl library installation. It is a good practice to keep them same.

Reply:
I don't think it's possible to specify CUDA 12.1 right now for PT, however it's the default. See https://pytorch.org/get-started/locally/

```
ERROR: Could not find a version that satisfies the requirement torch==2.1.2+cu121 (from versions: 1.7.1, 1.8.0, 1.8.1, 1.9.0, 1.9.1, 1.10.0, 1.10.1, 1.10.2, 1.11.0, 1.12.0, 1.12.1, 1.13.0, 1.13.1, 2.0.0, 2.0.1, 2.1.0, 2.1.1, 2.1.2, 2.2.0, 2.2.1, 2.2.2)
#8 5.320 ERROR: No matching distribution found for torch==2.1.2+cu121
```

For 11.8 we would need to use `pip3 install torch torchvision torchaudio --index-url https://download.pytorch.org/whl/cu118`

Reply:
I see. Then we can leave if for now.